### ISSUE TITLE
Transip Stack Storage Support 

### ISSUE THREAD

**Issue by mehargags** [rclone/rclone#3126]
It would be great to have support for Storage Stack provided by https://www.transip.nl/stack/
Looks like they run it on top of Owncloud
https://blog.caroga.net/how-to-synchronize-your-files-with-transips-stack-using-the-commandline/
not sure if your Owncloud support would work ? I Tried setting up with Webdav (using Access token) but it won't list any of my folders inside it, it successfully lists the files in the root as well as /DIR name as well.

Secondly "rclone sync" doesn't see to put in any files into the stack storage... what would be the issue ?

I found some API
https://www.transip.nl/knowledgebase/tag/64-api/
https://pypi.org/project/transip-stack-api/1.0.0/#files

**Comment by ncw** [comment]
Getting it to work with WebDAV would probably be the easiest thing to do.

Can you get free accounts?  Or could you lend me the credentials for yours for a bit?

**Comment by stefaang** [comment]
I've just tested rclone with my account.

```
[stack]
type = webdav
url = https://someuser.stackstorage.com/remote.php/webdav
vendor = owncloud
user = stefaang
pass = encrypted
```

I'm able to list, copy, sync. The second sync doesn't skip any files though.

```
stefaan@host:~/isos$ rclone sync somedir stack:somedir --progress
Transferred:       18.999M / 18.999 MBytes, 100%, 1.440 MBytes/s, ETA 0s
Errors:                 0
Checks:                28 / 28, 100%
Transferred:           19 / 19, 100%
Elapsed time:       13.1s
stefaan@host:~/isos$ rclone sync somedir stack:somedir --progress
Transferred:       18.999M / 18.999 MBytes, 100%, 1.799 MBytes/s, ETA 0s
Errors:                 0
Checks:                19 / 19, 100%
Transferred:           19 / 19, 100%
Elapsed time:       10.5s

```

**Comment by ncw** [comment]
I see why it is trying to upload everything twice...

You have the "owncloud" type configured, and rclone is trying to upload the files with timestamps

```
2019/04/25 16:51:06 DEBUG : PUT /remote.php/webdav/jikumel HTTP/1.1
Host: tafelpoowt.stackstorage.com
User-Agent: rclone/v1.47.0-022-ge275ad80-beta
Content-Length: 73
Authorization: XXXX
Content-Type: application/octet-stream
Oc-Checksum: SHA1:cf3b1d3c67da30ed95a50e4dcb69dbf16262a7d9
X-Oc-Mtime: 1556207416.529042
Accept-Encoding: gzip
```

However the timestamps of the files are not being updated, they are being set to the time of upload.

I don't know why the server doesn't support `X-Oc-Mtime` that might be something to ask the provider. It doesn't seem to support `Oc-Checksum` either.

If you set `vendor = other` then the sync will work properly, but losing the modification times and the sha1sums is not ideal, so might be worth a ticket to the vendor. They might be filtering them out in their frontent proxies or something like that.

> The webdav token doesn't work.
> Token: qA0ekrjw281EJKPPLQHLUf4oaX8

How have you been attempting to use that?

**Comment by stefaang** [comment]
In the [Settings](https://tafelpoowt.stackstorage.com/settings) tab you can create access tokens for webdav.
One would assume this could replace the user / password config... but it doesn't.

**Comment by stefaang** [comment]
I contacted support and they said the access tokens are related to 2FA.
When 2FA is not enabled, you need to stick to password authentication.

They will look into the issue with the headers.

**Comment by ncw** [comment]
> I contacted support and they said the access tokens are related to 2FA.
> When 2FA is not enabled, you need to stick to password authentication.
> 
> They will look into the issue with the headers.

Great - let me know what happens.

**Comment by lisxray** [comment]
> I contacted support and they said the access tokens are related to 2FA.
> When 2FA is not enabled, you need to stick to password authentication.
> 
> They will look into the issue with the headers.

Any update?😊

**Comment by vansante** [comment]
In what format does RClone send the `X-Oc-Mtime` header? STACK only supports an integer with a unix timestamp currently. It seems you are sending a number with microseconds which cannot be parsed and is silently ignored.

**Comment by leander091** [comment]
It looks indeed that the format of the X-Oc-Mtime header is invalid. I've done some troubleshooting with the official Stack client and if you upload a file to your storage the header will be like the included image. The Mtime is in the format of a Unix timestamp without any nano seconds.

![image](https://user-images.githubusercontent.com/5831104/73777456-60e20580-4781-11ea-9c9f-8d462c9c50ac.png)
 i guess the code below should be edited but i haven't tested it yet.

`	if f.useOCMtime {
		opts.ExtraHeaders["X-OC-Mtime"] = fmt.Sprintf("%f",float64(src.ModTime(ctx).UnixNano())/1e9)
	}`

**Comment by ncw** [comment]
Here is an rclone with X-OC-Mtime as an integer - can you give it a test?

https://beta.rclone.org/branch/v1.51.0-009-g629a88d5-fix-3126-webdav-time-beta/ (uploaded in 15-30 mins)

I can't find the specification for X-OC-Mtime - can anyone point me at them?

Certainly owncloud and nextcloud only keep timestamps to the nearest second so I'm not sure why I ever thought it needed fractional seconds.

**Comment by leander091** [comment]
Ive tested the new version but I saw still that the Mtime header is in an incorrect format.  The fractional seconds are not included any more but they are replaced with zero's. So the application discards it in silence.

 i have not found a format specification about the header on either the Nextcloud or Owncloud documentation. But the is an direction on the following page:  [https://docs.nextcloud.com/server/18/developer_manual/client_apis/WebDAV/chunking.html?highlight=mtime](url)

So in my understanding the X-OC-Mtime is just an simple representation of an Unix timestamp returned with the default PHP functions (getMtime()) and formatted without any changes in Next or owncloud.

**Comment by ncw** [comment]
@leander091 I made a mistake with that one try this!


https://beta.rclone.org/branch/v1.51.0-009-gdfc31ab7-fix-3126-webdav-time-beta/ (uploaded in 15-30 mins)

> I have not found a format specification about the header on either the Nextcloud or Owncloud documentation. But the is an direction on the following page: [https://docs.nextcloud.com/server/18/developer_manual/client_apis/WebDAV/chunking.html?highlight=mtime](url)
> 
> So in my understanding the X-OC-Mtime is just an simple representation of an Unix timestamp returned with the default PHP functions (getMtime()) and formatted without any changes in Next or owncloud.

That certainly looks suggestive that it is an integer...

nextcloud and owncloud are ignoring the extra precision too.

**Comment by leander091** [comment]
@ncw  I have tested the new beta and it works like a charm.  It checks the modification time as expected and does not upload a file if the time is equal to the version on the server.

**Comment by ncw** [comment]
Thanks for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52

**Comment by jaap** [comment]
Great! I can validate it working by using a bearer-token.
